fineract-provider/src/main/java/org/apache/fineract/portfolio/client/service/ClientWritePlatformServiceJpaRepositoryImpl.java

```diff
@@ -844,4 +844,59 @@ public CommandProcessingResult reActivateClient(Long entityId, JsonCommand comma
                 .withEntityId(entityId) //
                 .build();
     }
-}
\ No newline at end of file
+
+	@Override
+	public CommandProcessingResult undoRejection(Long entityId, JsonCommand command) {
+		final AppUser currentUser = this.context.authenticatedUser();
+		this.fromApiJsonDeserializer.validateUndoRejection(command);
+
+		final Client client = this.clientRepository.findOneWithNotFoundDetection(entityId);
+		final LocalDate undoRejectDate = command
+				.localDateValueOfParameterNamed(ClientApiConstants.reopenedDateParamName);
+
+		if (!client.isRejected()) {
+			final String errorMessage = "only rejected clients may be reactivated.";
+			throw new InvalidClientStateTransitionException("undorejection", "on.nonrejected.account", errorMessage);
+		} else if (client.getRejectedDate().isAfter(undoRejectDate)) {
+			final String errorMessage = "The client reactivation date cannot be before the client rejected date.";
+			throw new InvalidClientStateTransitionException("reopened", "date.cannot.before.client.rejected.date",
+					errorMessage, undoRejectDate, client.getRejectedDate());
+		}
+
+		client.reOpened(currentUser, undoRejectDate.toDate());
+		this.clientRepository.saveAndFlush(client);
+
+		return new CommandProcessingResultBuilder() //
+				.withCommandId(command.commandId()) //
+				.withClientId(entityId) //
+				.withEntityId(entityId) //
+				.build();
+	}
+
+	@Override
+	public CommandProcessingResult undoWithdrawal(Long entityId, JsonCommand command) {
+		final AppUser currentUser = this.context.authenticatedUser();
+		this.fromApiJsonDeserializer.validateUndoWithDrawn(command);
+
+		final Client client = this.clientRepository.findOneWithNotFoundDetection(entityId);
+		final LocalDate undoWithdrawalDate = command
+				.localDateValueOfParameterNamed(ClientApiConstants.reopenedDateParamName);
+
+		if (!client.isWithdrawn()) {
```

Comment:
I do not think moving the validation to validator is absolutely required here. We are still avoiding entity creation, which is a good benefit.